### Loading dataframes

In [287]:
import pandas as pd

### Check for duplicates

In [288]:
df = pd.DataFrame({'a':[1,1,1,2,2,3,4,5],
                  'b':[10,10,11,20,20,30,40,50]})

In [289]:
df

a   b
0  1  10
1  1  10
2  1  11
3  2  20
4  2  20
5  3  30
6  4  40
7  5  50

In [290]:
df.duplicated()

0    False
1     True
2    False
3    False
4     True
5    False
6    False
7    False
dtype: bool

In [291]:
df.duplicated(subset=['a'])

0    False
1     True
2     True
3    False
4     True
5    False
6    False
7    False
dtype: bool

In [292]:
df.duplicated().sum()

2

### Drop duplicated rows

In [293]:
df_no_duplicates = df.drop_duplicates()

In [294]:
len(df), len(df_no_duplicates)

(8, 6)

In [295]:
df.drop_duplicates(inplace=True)

In [296]:
len(df)

6

### Check for missing values

In [297]:
df = pd.DataFrame({'a':[1,2,3,4],
                  'b':[10,None,30,40],
                  'c':[None,None,None,400]})

In [298]:
df.isnull()

a      b      c
0  False  False   True
1  False   True   True
2  False  False   True
3  False  False  False

In [299]:
df.isnull().any(axis=0) # column-wise

a    False
b     True
c     True
dtype: bool

In [300]:
df.isnull().any(axis=1) # row-wise

0     True
1     True
2     True
3    False
dtype: bool

In [301]:
df.isnull().sum()

a    0
b    1
c    3
dtype: int64

### Drop rows with missing values

In [302]:
df

a     b      c
0  1  10.0    NaN
1  2   NaN    NaN
2  3  30.0    NaN
3  4  40.0  400.0

In [303]:
df_only_full_rows = df.dropna()
df_only_full_rows

a     b      c
3  4  40.0  400.0

In [304]:
df_rows_where_b_is_not_missing = df.dropna(subset=['b'])
df_rows_where_b_is_not_missing

a     b      c
0  1  10.0    NaN
2  3  30.0    NaN
3  4  40.0  400.0

### Replace missing values

In [305]:
df

a     b      c
0  1  10.0    NaN
1  2   NaN    NaN
2  3  30.0    NaN
3  4  40.0  400.0

In [306]:
mean_b = df['b'].mean()
df_missing_b_replaced_with_mean = df.fillna(value={'b':mean_b})
df_missing_b_replaced_with_mean

a          b      c
0  1  10.000000    NaN
1  2  26.666667    NaN
2  3  30.000000    NaN
3  4  40.000000  400.0

In [307]:
df_missing_b_replaced_with_mean_missing_c_replaced_with_zero = df.fillna(value={'b':mean_b,
                                                                                'c':0})
df_missing_b_replaced_with_mean_missing_c_replaced_with_zero

a          b      c
0  1  10.000000    0.0
1  2  26.666667    0.0
2  3  30.000000    0.0
3  4  40.000000  400.0

In [308]:
df.fillna('Unknown', inplace=True)
df

a        b        c
0  1       10  Unknown
1  2  Unknown  Unknown
2  3       30  Unknown
3  4       40      400

### Describing numeric data

In [309]:
df = pd.DataFrame({'a':[1,1,1,2,2,3,4,5],
                  'b':[10,10,11,20,None,None,40,50],
                  'c':['apple','apple','plum','pear','plum','apple','apple','apple']})

In [310]:
df

a     b      c
0  1  10.0  apple
1  1  10.0  apple
2  1  11.0   plum
3  2  20.0   pear
4  2   NaN   plum
5  3   NaN  apple
6  4  40.0  apple
7  5  50.0  apple

In [311]:
df.dtypes

a      int64
b    float64
c     object
dtype: object

In [312]:
df['c'].dtype

dtype('O')

In [313]:
df.describe()

a          b
count  8.000000   6.000000
mean   2.375000  23.500000
std    1.505941  17.363755
min    1.000000  10.000000
25%    1.000000  10.250000
50%    2.000000  15.500000
75%    3.250000  35.000000
max    5.000000  50.000000

In [314]:
df.max()

a       5
b      50
c    plum
dtype: object

In [315]:
df['a'].max()

5

### Describing non numeric data

In [316]:
df

a     b      c
0  1  10.0  apple
1  1  10.0  apple
2  1  11.0   plum
3  2  20.0   pear
4  2   NaN   plum
5  3   NaN  apple
6  4  40.0  apple
7  5  50.0  apple

In [317]:
df['c'].unique()

array(['apple', 'plum', 'pear'], dtype=object)

In [318]:
df['c'].nunique()

3

In [319]:
df['c'].value_counts()

apple    5
plum     2
pear     1
Name: c, dtype: int64

## Exercise

### 1 - exercise
Load the datasets into pandas dataframes called trip, weather and station. (Don't forget to import the pandas library first!) <br>
Create three variables called trip_duplicates_num, weather_duplicates_num and station_duplicatess_num that contains how many duplicated rows are in each dataframe.

In [320]:
### Your code here
import pandas as pd

class2_dir = 'C:\\Users\\szige\\Desktop\\CEU\\2019-2020 Fall\\Coding 3 - Data Management and Analysis with Python\\class2\\'
trip = pd.read_csv(class2_dir + 'trip.csv')
weather = pd.read_csv(class2_dir + 'weather.csv')
station = pd.read_csv(class2_dir + 'station.csv')

trip_duplicates_num = trip.duplicated().sum()
weather_duplicates_num = weather.duplicated().sum()
station_duplicates_num = station.duplicated().sum()

### 1 - check yourself

In [321]:
print('Length of trip dataframe should be 144115 with 100 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(trip), trip_duplicates_num))
print('Length of weather dataframe should be 928 with 8 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(weather), weather_duplicates_num))
print('Length of station dataframe should be 69 with 0 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(station), station_duplicates_num))

Length of trip dataframe should be 144115 with 100 duplicated rows.
The length of your dataframe is 144115 and you counted 100 duplicated rows

Length of weather dataframe should be 928 with 8 duplicated rows.
The length of your dataframe is 928 and you counted 8 duplicated rows

Length of station dataframe should be 69 with 0 duplicated rows.
The length of your dataframe is 69 and you counted 0 duplicated rows



### 2 - exercise
For all 3 dataframes delete the duplicated rows in place (without creating a new dataframe) <br>

In [322]:
### Your code here
trip.drop_duplicates(inplace=True)
weather.drop_duplicates(inplace=True)
station.drop_duplicates(inplace=True)

### 2 - check yourself

In [323]:
print('Length of trip dataframe should be 144015.\n\
The length of your dataframe is {}\n'.format(len(trip)))
print('Length of weather dataframe should be 920.\n\
The length of your dataframe is {}\n'.format(len(weather)))
print('Length of station dataframe should be 69.\n\
The length of your dataframe is {}\n'.format(len(station)))

Length of trip dataframe should be 144015.
The length of your dataframe is 144015

Length of weather dataframe should be 920.
The length of your dataframe is 920

Length of station dataframe should be 69.
The length of your dataframe is 69



### 3 - exercise
For all dataframes check if there are columns with missing values. <br>
Create 3 lists called trip_columns_with_missing_data, weather_columns_with_missing_data, station_columns_with_missing_data that contains the names of the columns with missing values of the dataframe. <br>
You can populate these lists by hand, or as an advanced task you can use pandas methods.

In [324]:
### Your code here
trip_columns_with_missing_data = trip.columns[trip.isnull().any()].tolist()
weather_columns_with_missing_data = weather.columns[weather.isnull().any()].tolist()
station_columns_with_missing_data = station.columns[station.isnull().any()].tolist()

### 3 - check yourself

In [325]:
print('Columns with missing values in the trip dataframe are:\n \
Subscription Type\n\
You have found:\n {}\n'.format(','.join(trip_columns_with_missing_data)))

print('Columns with missing values in the weather dataframe are:\n \
Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events\n\
You have found:\n {}\n'.format(','.join(weather_columns_with_missing_data)))

print('Columns with missing values in the station dataframe are:\n \
\n\
You have found:\n {}\n'.format(','.join(station_columns_with_missing_data)))

Columns with missing values in the trip dataframe are:
 Subscription Type
You have found:
 Subscription Type

Columns with missing values in the weather dataframe are:
 Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events
You have found:
 Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events

Columns with missing values in the station dataframe are:
 
You have found:
 



### 4 - exercise
How many values are missing in each column with missing values in the dataframes? <br>
Display the answer in any format you would like. As an advance task, try to display only the names of the columns that have missing values in them.

In [326]:
### Your code here
for c in trip_columns_with_missing_data:
    print('The number of missing values in ' + c + ' is ' + str(trip[c].isnull().sum()))

for c in weather_columns_with_missing_data:
    print('The number of missing values in ' + c + ' is ' + str(weather[c].isnull().sum()))
    
for c in station_columns_with_missing_data:
    print('The number of missing values in ' + c + ' is ' + str(station[c].isnull().sum()))

The number of missing values in Subscription Type is 10
The number of missing values in Max_Temperature_F is 3
The number of missing values in Mean_Temperature_F is 3
The number of missing values in Min_TemperatureF is 3
The number of missing values in Max_Gust_Speed_MPH is 138
The number of missing values in Events is 782


### 4 - check yourself
The number of missing values in Max_Temperature_F column is 3 <br>
The number of missing values in Mean_Temperature_F column is 3<br>
The number of missing values in Min_TemperatureF column is 3<br>
The number of missing values in Max_Gust_Speed_MPH column is 138<br>
The number of missing values in Events column is 782<br>
The number of missing values in Subscription Type column is 10<br>

### 5 - exercise
Before deciding how to deal with the missing values, let's get more familiar with the data! <br>
 - How many values are in the columns with missing data? As an advanced task, try to display the number of rows for only these columns, not the others.
 - Display the mean of each numeric column in the weather dataframe. Which are 2 columns with the lowest mean? As an advanced task, try to display them in descending order!
 - And what about the Events column in the weather dataframe? What are the unique values and how many times do they occur?

In [327]:
### Your code here
for c in trip_columns_with_missing_data:
    print(c + ' column is ' + str(len(trip[c])-trip[c].isnull().sum()))

for c in weather_columns_with_missing_data:
    print(c + ' column is ' + str(len(weather[c])-weather[c].isnull().sum()))
    
for c in station_columns_with_missing_data:
    print(c + ' column is ' + str(len(station[c])-station[c].isnull().sum()))

Subscription Type column is 144005
Max_Temperature_F column is 917
Mean_Temperature_F column is 917
Min_TemperatureF column is 917
Max_Gust_Speed_MPH column is 782
Events column is 138


In [328]:
weather.mean().sort_values(ascending=True)[:2]

Precipitation_In     0.021120
Cloud_Cover          2.143478
dtype: float64

In [329]:
weather['Events'].value_counts()

Rain        101
Fog          34
rain          2
Fog-Rain      1
Name: Events, dtype: int64

### 5 - check yourself
The number of rows without missing data are:
- Max_Temperature_F column is 917
- Mean_Temperature_F column is 917
- Min_TemperatureF column is 917
- Max_Gust_Speed_MPH column is 782
- Events column is 138
- Subscription Type column is 144005

The columns with the lowest mean values are Cloud_clover and Mean_Wind_Speed_MPH <br><br>
In the Events column there are 101 rows with Rain, 34 rows with Fog, 2 rows with rain and 1 row with Fog-Rain

### 6 - exercise
So let's decide what we will do with the missing data!<br>
- In the Temperature columns, there are not too many missing data, so let's fill those cells with the mean of the column.
- In the Gust Speed column there are more than 10% of data missing so using the mean would be risky. We'll leave the missing values for now, and fill them later when using this dataframe.
- In the Events columns, the missing value means that there were no rain or fog that day. So let's fill those cells with the string 'no_event'
- In the Subscription type column, only a few rows are missing and we can't guess the original values. Let's delete those rows from the dataframe.

Create a new weather and trip dataframe called weather_filled and trip_filled where these solutions are applied!

In [330]:
### Your code here
weather_filled = weather.fillna(value={'Max_Temperature_F':weather['Max_Temperature_F'].mean(),
                     'Mean_Temperature_F':weather['Mean_Temperature_F'].mean(),
                     'Min_TemperatureF':weather['Min_TemperatureF'].mean(),
                                      'Events':'no_event'})
trip_filled = trip.dropna(subset=['Subscription Type'])

### 6 - check yourself

In [331]:
print('In the weather_filled dataframe the number of missing data should be 1\n \
and in your dataframe the number is {}\n'.format(weather_filled.isnull().any().sum()))
print('The number of rows where the replacing value is not correct in the weather_filled dataframe: \n \
in column Max_Temperature_F is {} \n \
in column Mean_Temperature_F is {} \n \
in column Min_TemperatureF is {} \n \
in column Events is {} \n '.format((weather_filled[weather.Max_Temperature_F.isnull()]['Max_Temperature_F'] != weather.Max_Temperature_F.mean()).sum(),
                                             (weather_filled[weather.Mean_Temperature_F.isnull()]['Mean_Temperature_F'] != weather.Mean_Temperature_F.mean()).sum(),
                                             (weather_filled[weather.Min_TemperatureF.isnull()]['Min_TemperatureF'] != weather.Min_TemperatureF.mean()).sum(),
                                             (weather_filled[weather.Events.isnull()]['Events'] != 'no_event').sum()))
print('The length of the trip_filled dataframe should be 144005\n \
and in your dataframe number is {}\n'.format(len(trip_filled)))



In the weather_filled dataframe the number of missing data should be 1
 and in your dataframe the number is 1

The number of rows where the replacing value is not correct in the weather_filled dataframe: 
 in column Max_Temperature_F is 0 
 in column Mean_Temperature_F is 0 
 in column Min_TemperatureF is 0 
 in column Events is 0 
 
The length of the trip_filled dataframe should be 144005
 and in your dataframe number is 144005



### 7 - exercise
Save these new dataframes into csv-s called weather_filled.csv and trip_filled.csv!

In [332]:
### Your code here
trip_filled.to_csv('trip_filled.csv', index=False)
weather_filled.to_csv('weather_filled.csv', index=False)

### 7 - check yourself

In [333]:
import os

if 'trip_filled.csv' in os.listdir():
    print('trip_filled.csv was successfully saved')
else:
    print('trip.csv was NOT successfully saved')
if 'weather_filled.csv' in os.listdir():
    print('weather_filled.csv was successfully saved')
else:
    print('weather_filled.csv was NOT successfully saved')

trip_filled.csv was successfully saved
weather_filled.csv was successfully saved


### +1 exercise

Let's repeat filling the missing values in the weather dataframe, but this time you'll first create the dictionary used in the values part of the fillna() function and fill it automatically instead of typing the names of the columns: <br>
 - Create an empty dictionary
 - Loop through the columns of the weather dataframe
 - For each numeric column (int or float) that has missing values in it, but in less than 10% of the rows, create a key in the dictionary and assign the mean of the column as the value
 - For each string column that has missing values in it, the value in the dictionary should be the string 'no_event'
 - Create a dataframe called weather_filled2 and apply the fillna function on it, with the dictionary you created
 - Check if the weather filled2 dataframe is the same as the weather_filled dataframe

In [334]:
import numpy as np

weather_dict = {}

for c in weather:
    if weather[c].isnull().sum() > 0:
        if weather[c].dtype == np.number:
            if weather[c].isnull().sum() / len(weather[c]) > 0.1:
                weather_dict.update({c:weather[c].mean()})
        else:
            weather_dict.update({c:'no_event'})

weather_filled2 = weather.fillna(value=weather_dict)

In [335]:
weather_filled.equals(weather_filled2)

False

In [336]:
weather_filled2

Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
0     8/29/2013               74.0                68.0              61.0   
1     8/30/2013               78.0                69.0              60.0   
2     8/31/2013               71.0                64.0              57.0   
3      9/1/2013               74.0                66.0              58.0   
4      9/2/2013               75.0                69.0              62.0   
..          ...                ...                 ...               ...   
923   11/6/2013               72.0                60.0              47.0   
924   11/7/2013               72.0                61.0              50.0   
925   11/8/2013               69.0                59.0              49.0   
926   11/9/2013               70.0                58.0              46.0   
927  11/10/2013               64.0                55.0              45.0   

     Max_Dew_Point_F  MeanDew_Point_F  Min_Dewpoint_F  Max_Humidity  \
0                 61               58              56            93   
1                 61               58              56            90   
2                 57               56              54            93   
3                 60               56              53            87   
4                 61               60              58            93   
..               ...              ...             ...           ...   
923               43               38              31            66   
924               52               45              38            86   
925               52               49              46            93   
926               51               47              43            93   
927               51               47              43            93   

     Mean_Humidity   Min_Humidity   ...  Mean_Visibility_Miles   \
0                75             57  ...                      10   
1                70             50  ...                      10   
2                75             57  ...                      10   
3                68             49  ...                      10   
4                77             61  ...                      10   
..              ...            ...  ...                     ...   
923              46             26  ...                      10   
924              59             31  ...                      10   
925              71             49  ...                       8   
926              67             40  ...                      10   
927              75             56  ...                       9   

     Min_Visibility_Miles   Max_Wind_Speed_MPH   Mean_Wind_Speed_MPH   \
0                       10                   23                    11   
1                        7                   29                    13   
2                       10                   26                    15   
3                       10                   25                    13   
4                        6                   23                    12   
..                     ...                  ...                   ...   
923                     10                    7                     2   
924                      8                   15                     3   
925                      2                   12                     4   
926                     10                   10                     2   
927                      5                   12                     2   

     Max_Gust_Speed_MPH  Precipitation_In   Cloud_Cover     Events  \
0                  28.0                0.0             4  no_event   
1                  35.0                0.0             2  no_event   
2                  31.0                0.0             4  no_event   
3                  29.0                0.0             4  no_event   
4                  30.0                0.0             6  no_event   
..                  ...                ...           ...       ...   
923                17.0                0.0             0  no_event   
924